lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,174 @@
+template<> struct combines_to_linear<Rotation, Eigen::Matrix3d> : std::true_type {};
+template<> struct combines_to_linear<Scaling, Rotation> : std::true_type {};
+template<> struct combines_to_linear<Scaling, Eigen::Matrix3d> : std::true_type {};
+template<> struct combines_to_linear<Eigen::Matrix3d, Rotation> : std::true_type {};
+template<> struct combines_to_linear<Eigen::Matrix3d, Scaling> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline 
+std::enable_if_t<combines_to_linear<T_LHS, T_RHS>::value, Eigen::Matrix3d> 
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+    return asEigenType(lhs) * asEigenType(rhs);
+}
+
+template<class T_LHS, class T_RHS> struct combines_to_affine : std::false_type {}; 
+template<> struct combines_to_affine<Rotation, Translation> : std::true_type {};
+template<> struct combines_to_affine<Rotation, Eigen::Affine3d> : std::true_type {};
+template<> struct combines_to_affine<Scaling, Translation> : std::true_type {};
+template<> struct combines_to_affine<Scaling, Eigen::Affine3d> : std::true_type {};
+template<> struct combines_to_affine<Eigen::Matrix3d, Translation> : std::true_type {};
+template<> struct combines_to_affine<Eigen::Affine3d, Rotation> : std::true_type {};
+template<> struct combines_to_affine<Eigen::Affine3d, Scaling> : std::true_type {};
+template<> struct combines_to_affine<Eigen::Affine3d, Translation> : std::true_type {};
+template<> struct combines_to_affine<Translation, Rotation> : std::true_type {};
+template<> struct combines_to_affine<Translation, Scaling> : std::true_type {};
+template<> struct combines_to_affine<Translation, Eigen::Matrix3d> : std::true_type {};
+template<> struct combines_to_affine<Translation, Eigen::Affine3d> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline 
+std::enable_if_t<combines_to_affine<T_LHS, T_RHS>::value, Eigen::Affine3d> 
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+    return Eigen::Affine3d(asEigenType(lhs) * asEigenType(rhs));
+}
+
+template<class T_LHS> struct applies_to_vector : std::false_type {}; 
+template<> struct applies_to_vector<Rotation> : std::true_type {};
+template<> struct applies_to_vector<Translation> : std::true_type {};
+template<> struct applies_to_vector<Scaling> : std::true_type {};
+
+template <typename T_LHS>
+[[nodiscard]] inline 
+std::enable_if_t<applies_to_vector<T_LHS>::value, Eigen::Vector3d> 
+operator*(const T_LHS &lhs, const Eigen::Vector3d &rhs) {
+    return asEigenType(lhs) * rhs;
+}
+
+[[nodiscard]] inline Rotation operator*(const Rotation &lhs,
+                                                 const Rotation &rhs) {
+  return Rotation(lhs.quat() * rhs.quat());
+};
+
+[[nodiscard]] inline Scaling operator*(const Scaling &lhs,
+                                                const Scaling &rhs) {
+  return Scaling(
+      (lhs.matrix() * rhs.matrix()).diagonal().asDiagonal());
+};
```

Comment:
It doesn't look like it to me - here is the error I get in this function if I try passing through `Scaling` as an `Eigen::DiagonalMatrix<...>` (as opposed to `Eigen::Matrix3d`) everywhere, and change this function to simply be `return Scaling(lhs.matrix() * rhs.matrix());`:

```
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): error C2678: binary '*': no operator found which takes a left-hand operand of type 'Eigen::DiagonalMatrix<double,3,3>' (or there is no acceptable conversion) (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(76,5): message : could be 'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::Matrix<double,3,1,0,3,1>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &) const' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(157,30): message : or       'scipp::core::Scaling scipp::core::operator *(const scipp::core::Scaling &,const scipp::core::Scaling &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(152,31): message : or       'scipp::core::Rotation scipp::core::operator *(const scipp::core::Rotation &,const scipp::core::Rotation &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(82,5): message : or       'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>,const Eigen::Matrix<double,3,1,0,3,1>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &,const Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>> &)' [found using argument-dependent lookup] (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): message : while trying to match the argument list '(Eigen::DiagonalMatrix<double,3,3>, Eigen::DiagonalMatrix<double,3,3>)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): error C2678: binary '*': no operator found which takes a left-hand operand of type 'Eigen::DiagonalMatrix<double,3,3>' (or there is no acceptable conversion) (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(76,5): message : could be 'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::Matrix<double,3,1,0,3,1>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &) const' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(157,30): message : or       'scipp::core::Scaling scipp::core::operator *(const scipp::core::Scaling &,const scipp::core::Scaling &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(152,31): message : or       'scipp::core::Rotation scipp::core::operator *(const scipp::core::Rotation &,const scipp::core::Rotation &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(82,5): message : or       'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>,const Eigen::Matrix<double,3,1,0,3,1>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &,const Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>> &)' [found using argument-dependent lookup] (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): message : while trying to match the argument list '(Eigen::DiagonalMatrix<double,3,3>, Eigen::DiagonalMatrix<double,3,3>)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
```

I can get around that by later adding a bunch of `static_cast`s back to `Matrix3d` in various places but that does not feel like the right approach at all.

Reply:
Then maybe we should just multiply the underlying diagonal terms?

Reply:
Maybe using Eigen internally is pointless in that case, and we should consider implementing our `Scaling` class directly?

Reply:
That would work for this specific function, but after this one if fixed (either by the approach you suggest or by explicit casts) then we run into similar problems in the templated `operator*` functions above, and it's not so clear what we would do there...

The problem is a little bit different because `Quat * DiagonalMatrix` in eigen is *defined*, but gives `Affine3d` rather than `Matrix3d`, but we need it to give `Matrix3d` because rotations/scaling/matrix won't have a unit associated (whereas translation and affine will). This can probably be made to work with enough template special cases and so on, but it seems much clearer to pass through the scaling as a matrix which makes eigen do the right thing in the first place?

Reply:
Maybe we should consider dropping the separate "scaling" dtype altogether, and just use a matrix (or a double, if uniform) for this from the start?

Reply:
We could do that, but it means 3x the memory use compared to storing it as a diagonalmatrix and only converting it to matrix as needed? Perhaps we aren't worried about that?

I guess from the user API perspective we would still have some constructors for scaling (to be consistent with translation/rotation), but they would just return `dtype=matrix` as opposed to `dtype=scaling`?

Reply:
> We could do that, but it means 3x the memory use compared to storing it as a diagonalmatrix and only converting it to matrix as needed? Perhaps we aren't worried about that?

I think this is something we can worry about when it turns out to be a problem? There may be very limited use of this dtype, so if it avoids some mess for now that would be the better choice. It is always easier to add something later than to remove it.

> I guess from the user API perspective we would still have some constructors for scaling (to be consistent with translation/rotation), but they would just return dtype=matrix as opposed to dtype=scaling?

Yes, I agree with that suggestion, similar to how we have `from_rotvec`.